### PR TITLE
[RLlib] APPO Training iteration fn.

### PR DESCRIPTION
**Pull request by sven1977**
APPO Training iteration fn.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/ppo/appo.py [hunk 1]
```diff
@@ -163,45 +160,36 @@ def training(
...
         self.workers.local_worker().foreach_policy_to_train(
             lambda p, _: p.update_target()
         )
 
+    def update_target_and_kl(self, train_results):
+        cur_ts = self._counters[STEPS_SAMPLED_COUNTER]
+        last_update = self._counters[LAST_TARGET_UPDATE_TS]
+        target_update_freq = self.config["num_sgd_iter"] * self.config[
+            "minibatch_buffer_size"]
+        if cur_ts - last_update > target_update_freq:
```
Comment (avnishn) [inline rllib/agents/ppo/appo.py:180]: Could we write this by instead subclassing the trainer of IMPALA? 

I feel like that would ready nicely and show the power of the training iteration fn

**Review by avnishn — COMMENTED** [review]
I have 1 question but otherwise lgtm

File: rllib/agents/ppo/appo.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/ppo/appo.py:196]: Yeah, you are right. I don't like this config["after_train_step"] hack either :)

Will do.

Comment (sven1977) [reply rllib/agents/ppo/appo.py:195]: done

**Comment by avnishn** [comment]
This is all looking pretty good, although is it learning?

**Comment by sven1977** [comment]
Hey @avnishn , please give this another go. I confirmed learning Pong is still good, e.g. on only 7 workers:

```
(base) ray@ip-172-31-83-222:~/riot_games_atari_benchmarks/ray$ tail ~/output.txt 
Resources requested: 0/8 CPUs, 0/1 GPUs, 0.0/35.56 GiB heap, 0.0/17.78 GiB objects (0.0/1.0 accelerator_type:V100)
Result logdir: /home/ray/ray_results/pong-appo
Number of trials: 1/1 (1 TERMINATED)
+-------------------------------------+------------+---------------------+--------+------------------+---------+----------+----------------------+----------------------+--------------------+
| Trial name                          | status     | loc                 |   iter |   total time (s) |      ts |   reward |   episode_reward_max |   episode_reward_min |   episode_len_mean |
|-------------------------------------+------------+---------------------+--------+------------------+---------+----------+----------------------+----------------------+--------------------|
| APPO_PongNoFrameskip-v4_7fb9b_00000 | TERMINATED | 172.31.83.222:15223 |    154 |          1697.41 | 2931200 |    18.07 |                   20 |                    8 |            7679.47 |
+-------------------------------------+------------+---------------------+--------+------------------+---------+----------+----------------------+----------------------+--------------------+
```

**Review by avnishn — APPROVED** [review]

**Pull request merged**